frontend/src/test/java/com/google/sps/InMemoryPlaceVisitStorageTest.java

```diff
@@ -0,0 +1,333 @@
+package com.google.sps;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.common.truth.Truth8.assertThat;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import java.util.List;
+import com.google.tripmeout.frontend.storage.InMemoryPlaceVisitStorage;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+
+@RunWith(JUnit4.class)
+public final class InMemoryPlaceVisitStorageTest {
+  private static final InMemoryPlaceVisitStorage NO_PLACE_VISITS = new InMemoryPlaceVisitStorage();
+
+  // PlaceVisitModel objects to use in testing
+  private static final PlaceVisitModel LONDON = PlaceVisitModel.builder()
+    .setPlaceId("LCY")
+    .setName("London, UK")
+    .setTripId("a")
+    .setUserMark("don't-care")
+    .setLatitude(56)
+    .setLongitude(23.64)
+    .build();
+
+  private static final PlaceVisitModel ROME = PlaceVisitModel.builder()
+    .setPlaceId("FCO")
+    .setName("Rome, Italy")
+    .setTripId("a")
+    .setUserMark("must-see")
+    .setLatitude(44.32)
+    .setLongitude(32.1244)
+    .build();
+
+  private static final PlaceVisitModel PARIS = PlaceVisitModel.builder()
+    .setPlaceId("CDG")
+    .setName("Paris, France")
+    .setTripId("a")
+    .setUserMark("if-time")
+    .setLatitude(48.3288)
+    .setLongitude(34)
+    .build();
+
+  private static final PlaceVisitModel TOKYO = PlaceVisitModel.builder()
+    .setPlaceId("HND")
+    .setName("Tokyo, Japan")
+    .setTripId("a")
+    .setUserMark("must-see")
+    .setLatitude(35.32)
+    .setLongitude(139.33)
+    .build();
+
+  private static final PlaceVisitModel BEIJING = PlaceVisitModel.builder()
+    .setPlaceId("PEK")
+    .setName("Beijing, China")
+    .setTripId("b")
+    .setUserMark("must-see")
+    .setLatitude(35.32)
+    .setLongitude(125.33)
+    .build();
+
+  private static final PlaceVisitModel SEOUL = PlaceVisitModel.builder()
+    .setPlaceId("ICN")
+    .setName("Seoul, South Korea")
+    .setTripId("b")
+    .setUserMark("if-time")
+    .setLatitude(26.32)
+    .setLongitude(135.33)
+    .build();
+
+  private static final PlaceVisitModel TOKYO_B = PlaceVisitModel.builder()
+    .setPlaceId("HND")
+    .setName("Tokyo, Japan")
+    .setTripId("b")
+    .setUserMark("must-see")
+    .setLatitude(35.32)
+    .setLongitude(139.33)
+    .build();
+
+  private static final PlaceVisitModel SEOUL_B = PlaceVisitModel.builder()
+    .setPlaceId("ICN")
+    .setName("some city in Korea")
+    .setTripId("b")
+    .setUserMark("if-time")
+    .setLatitude(0)
+    .setLongitude(0)
+    .build();
+
+  /** 
+   * test that a PlaceVisitAlreadyExistsException is thrown when a place that 
+   * has already been added to the storage is added again 
+   */
+  @Test(expected = PlaceVisitAlreadyExistsException.class)
```

Comment:
Using `@Test(expected = ...)` is usually the wrong thing. For example if suddenly the first `addPlaceVisit` or the constructor started throwing this, this test might pass. While other tests would hopefully start failing, it might make it more difficult to root cause the error.

Instead prefer to use `org.junit.Assert.assertThrows` as so:
```java
assertThrows(PlaceVisitAlreadyExistsException.class, () -> storage.addPlaceVisit(LONDON));
```

This will also return the exception to you if you want to assert other things about it like the cause, or that a specific string was included in the message for helpfulness.